File: 백준/Gold/2294. 동전 2/동전 2.py
import sys

input = sys.stdin.readline
N, K = map(int, input().split())
coins = [int(input()) for _ in range(N)]
dp = [0] * (K + 1)


def solution():
    for coin in coins:
        if coin < K + 1:
            dp[coin] = 1

    for i in range(1, K + 1):
        if not i in coins:
            cases = []
            for coin in coins:
                if -1 < i - coin and dp[i - coin] != 0:
                    cases.append(dp[i - coin])
            if cases:
                dp[i] = min(cases) + 1

    print(-1 if dp[K] == 0 else dp[K])


solution()

File: 백준/Gold/2294. 동전 2/동전 2.js
const fs = require('fs');
const input = fs.readFileSync(0, 'utf8').trim().split('\n');
const [n, k] = input[0].split(' ').map((v) => +v);
const coins = input.slice(1).map((v) => +v);

function solution(n, k, coins) {
  const dp = new Array(k + 1).fill(-1);
  dp[0] = 0;

  for (let i = 1; i <= k; i++) {
    const r = [];

    coins.forEach((coin) => {
      if (i - coin >= 0 && dp[i - coin] !== -1) r.push(dp[i - coin]);
    });

    if (r.length) dp[i] = Math.min(...r) + 1;
  }

  return dp[k];
}

console.log(solution(n, k, coins));
